### PR TITLE
Fix hash_lock / keystore.sk_dk_lock lock inversion

### PR DESCRIPTION
**Pull request by behlendorf**
### Description

The keystore.sk_dk_lock should not be held while performing I/O.  Drop the lock when reading from disk and update the code so they the first successful caller adds the key.

### Motivation and Context

#7112 

### How Has This Been Tested?

Locally built, overnight ztest run underway.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: module/zfs/dsl_crypt.c [hunk 1]
```diff
@@ -624,55 +624,56 @@ spa_keystore_dsl_key_hold_dd(spa_t *spa, dsl_dir_t *dd, void *tag,
...
 
-	rw_enter(&spa->spa_keystore.sk_dk_lock, RW_WRITER);
-
-	/* lookup the key in the tree of currently loaded keys */
-	ret = spa_keystore_dsl_key_hold_impl(spa, dckobj, tag, &dck);
+	/* Lookup the key in the tree of currently loaded keys */
+	rw_enter(&spa->spa_keystore.sk_dk_lock, RW_READER);
+	ret = spa_keystore_dsl_key_hold_impl(spa, dckobj, tag, &dck_ks);
+	rw_exit(&spa->spa_keystore.sk_dk_lock);
 	if (!ret) {
```
Comment (tcaputi) [inline module/zfs/dsl_crypt.c]: probably should change this to ret == 0 while we're here.

Comment (behlendorf) [reply module/zfs/dsl_crypt.c]: will do

**Comment by behlendorf** [comment]
@tcaputi I've refresh this can you look at it again.

**Comment by sempervictus** [comment]
Zlooped for the last ~4h, no sudden death yet.

**Comment by mailinglists35** [comment]
is this going to be included in 0.7.6 release?

**Comment by tcaputi** [comment]
> is this going to be included in 0.7.6 release?

Encryption is not going to be included at all until 0.8.0.

**Pull request merged**